// @babel/preset-env/lib/polyfills/regenerator/dist/entry-plugin.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = _default;

var _utils = require("../../utils");

function isRegeneratorSource(source) {
  return source === "regenerator-runtime/runtime";
}

function _default() {
  var visitor = {
    ImportDeclaration: function ImportDeclaration(path) {
      if (isRegeneratorSource((0, _utils.getImportSource)(path))) {
        this.regeneratorImportExcluded = true;
        path.remove();
      }
    },
    Program: function Program(path) {
      var _this = this;

      path.get("body").forEach(function (bodyPath) {
        if (isRegeneratorSource((0, _utils.getRequireSource)(bodyPath))) {
          _this.regeneratorImportExcluded = true;
          bodyPath.remove();
        }
      });
    }
  };
  return {
    name: "regenerator-entry",
    visitor: visitor,
    pre: function pre() {
      this.regeneratorImportExcluded = false;
    },
    post: function post() {
      if (this.opts.debug && this.regeneratorImportExcluded) {
        var filename = this.file.opts.filename;

        if (process.env.BABEL_ENV === "test") {
          filename = filename.replace(/\\/g, "/");
        }

        console.log("\n[".concat(filename, "] Based on your targets, regenerator-runtime import excluded."));
      }
    }
  };
}
