// enhanced-resolve/lib/dist/AppendPlugin.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

module.exports =
/*#__PURE__*/
function () {
  function AppendPlugin(source, appending, target) {
    _classCallCheck(this, AppendPlugin);

    this.source = source;
    this.appending = appending;
    this.target = target;
  }

  _createClass(AppendPlugin, [{
    key: "apply",
    value: function apply(resolver) {
      var _this = this;

      var target = resolver.ensureHook(this.target);
      resolver.getHook(this.source).tapAsync("AppendPlugin", function (request, resolveContext, callback) {
        var obj = Object.assign({}, request, {
          path: request.path + _this.appending,
          relativePath: request.relativePath && request.relativePath + _this.appending
        });
        resolver.doResolve(target, obj, _this.appending, resolveContext, callback);
      });
    }
  }]);

  return AppendPlugin;
}();
