// enhanced-resolve/lib/dist/CloneBasenamePlugin.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var basename = require("./getPaths").basename;

module.exports =
/*#__PURE__*/
function () {
  function CloneBasenamePlugin(source, target) {
    _classCallCheck(this, CloneBasenamePlugin);

    this.source = source;
    this.target = target;
  }

  _createClass(CloneBasenamePlugin, [{
    key: "apply",
    value: function apply(resolver) {
      var target = resolver.ensureHook(this.target);
      resolver.getHook(this.source).tapAsync("CloneBasenamePlugin", function (request, resolveContext, callback) {
        var filename = basename(request.path);
        var filePath = resolver.join(request.path, filename);
        var obj = Object.assign({}, request, {
          path: filePath,
          relativePath: request.relativePath && resolver.join(request.relativePath, filename)
        });
        resolver.doResolve(target, obj, "using path: " + filePath, resolveContext, callback);
      });
    }
  }]);

  return CloneBasenamePlugin;
}();
