// enhanced-resolve/lib/dist/ConcordModulesPlugin.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var concord = require("./concord");

var DescriptionFileUtils = require("./DescriptionFileUtils");

var getInnerRequest = require("./getInnerRequest");

module.exports =
/*#__PURE__*/
function () {
  function ConcordModulesPlugin(source, options, target) {
    _classCallCheck(this, ConcordModulesPlugin);

    this.source = source;
    this.options = options;
    this.target = target;
  }

  _createClass(ConcordModulesPlugin, [{
    key: "apply",
    value: function apply(resolver) {
      var target = resolver.ensureHook(this.target);
      resolver.getHook(this.source).tapAsync("ConcordModulesPlugin", function (request, resolveContext, callback) {
        var innerRequest = getInnerRequest(resolver, request);
        if (!innerRequest) return callback();
        var concordField = DescriptionFileUtils.getField(request.descriptionFileData, "concord");
        if (!concordField) return callback();
        var data = concord.matchModule(request.context, concordField, innerRequest);
        if (data === innerRequest) return callback();
        if (data === undefined) return callback();

        if (data === false) {
          var ignoreObj = Object.assign({}, request, {
            path: false
          });
          return callback(null, ignoreObj);
        }

        var obj = Object.assign({}, request, {
          path: request.descriptionFileRoot,
          request: data
        });
        resolver.doResolve(target, obj, "aliased from description file " + request.descriptionFilePath + " with mapping '" + innerRequest + "' to '" + data + "'", resolveContext, function (err, result) {
          if (err) return callback(err); // Don't allow other aliasing or raw request

          if (result === undefined) return callback(null, null);
          callback(null, result);
        });
      });
    }
  }]);

  return ConcordModulesPlugin;
}();
