// enhanced-resolve/lib/dist/ConcordMainPlugin.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var path = require("path");

var concord = require("./concord");

var DescriptionFileUtils = require("./DescriptionFileUtils");

module.exports =
/*#__PURE__*/
function () {
  function ConcordMainPlugin(source, options, target) {
    _classCallCheck(this, ConcordMainPlugin);

    this.source = source;
    this.options = options;
    this.target = target;
  }

  _createClass(ConcordMainPlugin, [{
    key: "apply",
    value: function apply(resolver) {
      var target = resolver.ensureHook(this.target);
      resolver.getHook(this.source).tapAsync("ConcordMainPlugin", function (request, resolveContext, callback) {
        if (request.path !== request.descriptionFileRoot) return callback();
        var concordField = DescriptionFileUtils.getField(request.descriptionFileData, "concord");
        if (!concordField) return callback();
        var mainModule = concord.getMain(request.context, concordField);
        if (!mainModule) return callback();
        var obj = Object.assign({}, request, {
          request: mainModule
        });
        var filename = path.basename(request.descriptionFilePath);
        return resolver.doResolve(target, obj, "use " + mainModule + " from " + filename, resolveContext, callback);
      });
    }
  }]);

  return ConcordMainPlugin;
}();
