// enhanced-resolve/lib/dist/JoinRequestPlugin.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

module.exports =
/*#__PURE__*/
function () {
  function JoinRequestPlugin(source, target) {
    _classCallCheck(this, JoinRequestPlugin);

    this.source = source;
    this.target = target;
  }

  _createClass(JoinRequestPlugin, [{
    key: "apply",
    value: function apply(resolver) {
      var target = resolver.ensureHook(this.target);
      resolver.getHook(this.source).tapAsync("JoinRequestPlugin", function (request, resolveContext, callback) {
        var obj = Object.assign({}, request, {
          path: resolver.join(request.path, request.request),
          relativePath: request.relativePath && resolver.join(request.relativePath, request.request),
          request: undefined
        });
        resolver.doResolve(target, obj, null, resolveContext, callback);
      });
    }
  }]);

  return JoinRequestPlugin;
}();
