// enhanced-resolve/lib/dist/ModulesInRootPlugin.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

module.exports =
/*#__PURE__*/
function () {
  function ModulesInRootPlugin(source, path, target) {
    _classCallCheck(this, ModulesInRootPlugin);

    this.source = source;
    this.path = path;
    this.target = target;
  }

  _createClass(ModulesInRootPlugin, [{
    key: "apply",
    value: function apply(resolver) {
      var _this = this;

      var target = resolver.ensureHook(this.target);
      resolver.getHook(this.source).tapAsync("ModulesInRootPlugin", function (request, resolveContext, callback) {
        var obj = Object.assign({}, request, {
          path: _this.path,
          request: "./" + request.request
        });
        resolver.doResolve(target, obj, "looking for modules in " + _this.path, resolveContext, callback);
      });
    }
  }]);

  return ModulesInRootPlugin;
}();
