// ajv-keywords/keywords/dist/uniqueItemProperties.dev.js
'use strict';

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var SCALAR_TYPES = ['number', 'integer', 'string', 'boolean', 'null'];

module.exports = function defFunc(ajv) {
  defFunc.definition = {
    type: 'array',
    compile: function compile(keys, parentSchema, it) {
      var equal = it.util.equal;
      var scalar = getScalarKeys(keys, parentSchema);
      return function (data) {
        if (data.length > 1) {
          for (var k = 0; k < keys.length; k++) {
            var i,
                key = keys[k];

            if (scalar[k]) {
              var hash = {};

              for (i = data.length; i--;) {
                if (!data[i] || _typeof(data[i]) != 'object') continue;
                var prop = data[i][key];
                if (prop && _typeof(prop) == 'object') continue;
                if (typeof prop == 'string') prop = '"' + prop;
                if (hash[prop]) return false;
                hash[prop] = true;
              }
            } else {
              for (i = data.length; i--;) {
                if (!data[i] || _typeof(data[i]) != 'object') continue;

                for (var j = i; j--;) {
                  if (data[j] && _typeof(data[j]) == 'object' && equal(data[i][key], data[j][key])) return false;
                }
              }
            }
          }
        }

        return true;
      };
    },
    metaSchema: {
      type: 'array',
      items: {
        type: 'string'
      }
    }
  };
  ajv.addKeyword('uniqueItemProperties', defFunc.definition);
  return ajv;
};

function getScalarKeys(keys, schema) {
  return keys.map(function (key) {
    var properties = schema.items && schema.items.properties;
    var propType = properties && properties[key] && properties[key].type;
    return Array.isArray(propType) ? propType.indexOf('object') < 0 && propType.indexOf('array') < 0 : SCALAR_TYPES.indexOf(propType) >= 0;
  });
}
