// eslint/lib/rules/dist/index.dev.js
/**
 * @fileoverview Collects the built-in rules into a map structure so that they can be imported all at once and without
 * using the file-system directly.
 * @author Peter (Somogyvari) Metz
 */
"use strict";
/* eslint sort-keys: ["error", "asc"] */

var _require = require("./utils/lazy-loading-rule-map"),
    LazyLoadingRuleMap = _require.LazyLoadingRuleMap;
/** @type {Map<string, import("../shared/types").Rule>} */


module.exports = new LazyLoadingRuleMap(Object.entries({
  "accessor-pairs": function accessorPairs() {
    return require("./accessor-pairs");
  },
  "array-bracket-newline": function arrayBracketNewline() {
    return require("./array-bracket-newline");
  },
  "array-bracket-spacing": function arrayBracketSpacing() {
    return require("./array-bracket-spacing");
  },
  "array-callback-return": function arrayCallbackReturn() {
    return require("./array-callback-return");
  },
  "array-element-newline": function arrayElementNewline() {
    return require("./array-element-newline");
  },
  "arrow-body-style": function arrowBodyStyle() {
    return require("./arrow-body-style");
  },
  "arrow-parens": function arrowParens() {
    return require("./arrow-parens");
  },
  "arrow-spacing": function arrowSpacing() {
    return require("./arrow-spacing");
  },
  "block-scoped-var": function blockScopedVar() {
    return require("./block-scoped-var");
  },
  "block-spacing": function blockSpacing() {
    return require("./block-spacing");
  },
  "brace-style": function braceStyle() {
    return require("./brace-style");
  },
  "callback-return": function callbackReturn() {
    return require("./callback-return");
  },
  camelcase: function camelcase() {
    return require("./camelcase");
  },
  "capitalized-comments": function capitalizedComments() {
    return require("./capitalized-comments");
  },
  "class-methods-use-this": function classMethodsUseThis() {
    return require("./class-methods-use-this");
  },
  "comma-dangle": function commaDangle() {
    return require("./comma-dangle");
  },
  "comma-spacing": function commaSpacing() {
    return require("./comma-spacing");
  },
  "comma-style": function commaStyle() {
    return require("./comma-style");
  },
  complexity: function complexity() {
    return require("./complexity");
  },
  "computed-property-spacing": function computedPropertySpacing() {
    return require("./computed-property-spacing");
  },
  "consistent-return": function consistentReturn() {
    return require("./consistent-return");
  },
  "consistent-this": function consistentThis() {
    return require("./consistent-this");
  },
  "constructor-super": function constructorSuper() {
    return require("./constructor-super");
  },
  curly: function curly() {
    return require("./curly");
  },
  "default-case": function defaultCase() {
    return require("./default-case");
  },
  "default-param-last": function defaultParamLast() {
    return require("./default-param-last");
  },
  "dot-location": function dotLocation() {
    return require("./dot-location");
  },
  "dot-notation": function dotNotation() {
    return require("./dot-notation");
  },
  "eol-last": function eolLast() {
    return require("./eol-last");
  },
  eqeqeq: function eqeqeq() {
    return require("./eqeqeq");
  },
  "for-direction": function forDirection() {
    return require("./for-direction");
  },
  "func-call-spacing": function funcCallSpacing() {
    return require("./func-call-spacing");
  },
  "func-name-matching": function funcNameMatching() {
    return require("./func-name-matching");
  },
  "func-names": function funcNames() {
    return require("./func-names");
  },
  "func-style": function funcStyle() {
    return require("./func-style");
  },
  "function-call-argument-newline": function functionCallArgumentNewline() {
    return require("./function-call-argument-newline");
  },
  "function-paren-newline": function functionParenNewline() {
    return require("./function-paren-newline");
  },
  "generator-star-spacing": function generatorStarSpacing() {
    return require("./generator-star-spacing");
  },
  "getter-return": function getterReturn() {
    return require("./getter-return");
  },
  "global-require": function globalRequire() {
    return require("./global-require");
  },
  "grouped-accessor-pairs": function groupedAccessorPairs() {
    return require("./grouped-accessor-pairs");
  },
  "guard-for-in": function guardForIn() {
    return require("./guard-for-in");
  },
  "handle-callback-err": function handleCallbackErr() {
    return require("./handle-callback-err");
  },
  "id-blacklist": function idBlacklist() {
    return require("./id-blacklist");
  },
  "id-length": function idLength() {
    return require("./id-length");
  },
  "id-match": function idMatch() {
    return require("./id-match");
  },
  "implicit-arrow-linebreak": function implicitArrowLinebreak() {
    return require("./implicit-arrow-linebreak");
  },
  indent: function indent() {
    return require("./indent");
  },
  "indent-legacy": function indentLegacy() {
    return require("./indent-legacy");
  },
  "init-declarations": function initDeclarations() {
    return require("./init-declarations");
  },
  "jsx-quotes": function jsxQuotes() {
    return require("./jsx-quotes");
  },
  "key-spacing": function keySpacing() {
    return require("./key-spacing");
  },
  "keyword-spacing": function keywordSpacing() {
    return require("./keyword-spacing");
  },
  "line-comment-position": function lineCommentPosition() {
    return require("./line-comment-position");
  },
  "linebreak-style": function linebreakStyle() {
    return require("./linebreak-style");
  },
  "lines-around-comment": function linesAroundComment() {
    return require("./lines-around-comment");
  },
  "lines-around-directive": function linesAroundDirective() {
    return require("./lines-around-directive");
  },
  "lines-between-class-members": function linesBetweenClassMembers() {
    return require("./lines-between-class-members");
  },
  "max-classes-per-file": function maxClassesPerFile() {
    return require("./max-classes-per-file");
  },
  "max-depth": function maxDepth() {
    return require("./max-depth");
  },
  "max-len": function maxLen() {
    return require("./max-len");
  },
  "max-lines": function maxLines() {
    return require("./max-lines");
  },
  "max-lines-per-function": function maxLinesPerFunction() {
    return require("./max-lines-per-function");
  },
  "max-nested-callbacks": function maxNestedCallbacks() {
    return require("./max-nested-callbacks");
  },
  "max-params": function maxParams() {
    return require("./max-params");
  },
  "max-statements": function maxStatements() {
    return require("./max-statements");
  },
  "max-statements-per-line": function maxStatementsPerLine() {
    return require("./max-statements-per-line");
  },
  "multiline-comment-style": function multilineCommentStyle() {
    return require("./multiline-comment-style");
  },
  "multiline-ternary": function multilineTernary() {
    return require("./multiline-ternary");
  },
  "new-cap": function newCap() {
    return require("./new-cap");
  },
  "new-parens": function newParens() {
    return require("./new-parens");
  },
  "newline-after-var": function newlineAfterVar() {
    return require("./newline-after-var");
  },
  "newline-before-return": function newlineBeforeReturn() {
    return require("./newline-before-return");
  },
  "newline-per-chained-call": function newlinePerChainedCall() {
    return require("./newline-per-chained-call");
  },
  "no-alert": function noAlert() {
    return require("./no-alert");
  },
  "no-array-constructor": function noArrayConstructor() {
    return require("./no-array-constructor");
  },
  "no-async-promise-executor": function noAsyncPromiseExecutor() {
    return require("./no-async-promise-executor");
  },
  "no-await-in-loop": function noAwaitInLoop() {
    return require("./no-await-in-loop");
  },
  "no-bitwise": function noBitwise() {
    return require("./no-bitwise");
  },
  "no-buffer-constructor": function noBufferConstructor() {
    return require("./no-buffer-constructor");
  },
  "no-caller": function noCaller() {
    return require("./no-caller");
  },
  "no-case-declarations": function noCaseDeclarations() {
    return require("./no-case-declarations");
  },
  "no-catch-shadow": function noCatchShadow() {
    return require("./no-catch-shadow");
  },
  "no-class-assign": function noClassAssign() {
    return require("./no-class-assign");
  },
  "no-compare-neg-zero": function noCompareNegZero() {
    return require("./no-compare-neg-zero");
  },
  "no-cond-assign": function noCondAssign() {
    return require("./no-cond-assign");
  },
  "no-confusing-arrow": function noConfusingArrow() {
    return require("./no-confusing-arrow");
  },
  "no-console": function noConsole() {
    return require("./no-console");
  },
  "no-const-assign": function noConstAssign() {
    return require("./no-const-assign");
  },
  "no-constant-condition": function noConstantCondition() {
    return require("./no-constant-condition");
  },
  "no-constructor-return": function noConstructorReturn() {
    return require("./no-constructor-return");
  },
  "no-continue": function noContinue() {
    return require("./no-continue");
  },
  "no-control-regex": function noControlRegex() {
    return require("./no-control-regex");
  },
  "no-debugger": function noDebugger() {
    return require("./no-debugger");
  },
  "no-delete-var": function noDeleteVar() {
    return require("./no-delete-var");
  },
  "no-div-regex": function noDivRegex() {
    return require("./no-div-regex");
  },
  "no-dupe-args": function noDupeArgs() {
    return require("./no-dupe-args");
  },
  "no-dupe-class-members": function noDupeClassMembers() {
    return require("./no-dupe-class-members");
  },
  "no-dupe-else-if": function noDupeElseIf() {
    return require("./no-dupe-else-if");
  },
  "no-dupe-keys": function noDupeKeys() {
    return require("./no-dupe-keys");
  },
  "no-duplicate-case": function noDuplicateCase() {
    return require("./no-duplicate-case");
  },
  "no-duplicate-imports": function noDuplicateImports() {
    return require("./no-duplicate-imports");
  },
  "no-else-return": function noElseReturn() {
    return require("./no-else-return");
  },
  "no-empty": function noEmpty() {
    return require("./no-empty");
  },
  "no-empty-character-class": function noEmptyCharacterClass() {
    return require("./no-empty-character-class");
  },
  "no-empty-function": function noEmptyFunction() {
    return require("./no-empty-function");
  },
  "no-empty-pattern": function noEmptyPattern() {
    return require("./no-empty-pattern");
  },
  "no-eq-null": function noEqNull() {
    return require("./no-eq-null");
  },
  "no-eval": function noEval() {
    return require("./no-eval");
  },
  "no-ex-assign": function noExAssign() {
    return require("./no-ex-assign");
  },
  "no-extend-native": function noExtendNative() {
    return require("./no-extend-native");
  },
  "no-extra-bind": function noExtraBind() {
    return require("./no-extra-bind");
  },
  "no-extra-boolean-cast": function noExtraBooleanCast() {
    return require("./no-extra-boolean-cast");
  },
  "no-extra-label": function noExtraLabel() {
    return require("./no-extra-label");
  },
  "no-extra-parens": function noExtraParens() {
    return require("./no-extra-parens");
  },
  "no-extra-semi": function noExtraSemi() {
    return require("./no-extra-semi");
  },
  "no-fallthrough": function noFallthrough() {
    return require("./no-fallthrough");
  },
  "no-floating-decimal": function noFloatingDecimal() {
    return require("./no-floating-decimal");
  },
  "no-func-assign": function noFuncAssign() {
    return require("./no-func-assign");
  },
  "no-global-assign": function noGlobalAssign() {
    return require("./no-global-assign");
  },
  "no-implicit-coercion": function noImplicitCoercion() {
    return require("./no-implicit-coercion");
  },
  "no-implicit-globals": function noImplicitGlobals() {
    return require("./no-implicit-globals");
  },
  "no-implied-eval": function noImpliedEval() {
    return require("./no-implied-eval");
  },
  "no-import-assign": function noImportAssign() {
    return require("./no-import-assign");
  },
  "no-inline-comments": function noInlineComments() {
    return require("./no-inline-comments");
  },
  "no-inner-declarations": function noInnerDeclarations() {
    return require("./no-inner-declarations");
  },
  "no-invalid-regexp": function noInvalidRegexp() {
    return require("./no-invalid-regexp");
  },
  "no-invalid-this": function noInvalidThis() {
    return require("./no-invalid-this");
  },
  "no-irregular-whitespace": function noIrregularWhitespace() {
    return require("./no-irregular-whitespace");
  },
  "no-iterator": function noIterator() {
    return require("./no-iterator");
  },
  "no-label-var": function noLabelVar() {
    return require("./no-label-var");
  },
  "no-labels": function noLabels() {
    return require("./no-labels");
  },
  "no-lone-blocks": function noLoneBlocks() {
    return require("./no-lone-blocks");
  },
  "no-lonely-if": function noLonelyIf() {
    return require("./no-lonely-if");
  },
  "no-loop-func": function noLoopFunc() {
    return require("./no-loop-func");
  },
  "no-magic-numbers": function noMagicNumbers() {
    return require("./no-magic-numbers");
  },
  "no-misleading-character-class": function noMisleadingCharacterClass() {
    return require("./no-misleading-character-class");
  },
  "no-mixed-operators": function noMixedOperators() {
    return require("./no-mixed-operators");
  },
  "no-mixed-requires": function noMixedRequires() {
    return require("./no-mixed-requires");
  },
  "no-mixed-spaces-and-tabs": function noMixedSpacesAndTabs() {
    return require("./no-mixed-spaces-and-tabs");
  },
  "no-multi-assign": function noMultiAssign() {
    return require("./no-multi-assign");
  },
  "no-multi-spaces": function noMultiSpaces() {
    return require("./no-multi-spaces");
  },
  "no-multi-str": function noMultiStr() {
    return require("./no-multi-str");
  },
  "no-multiple-empty-lines": function noMultipleEmptyLines() {
    return require("./no-multiple-empty-lines");
  },
  "no-native-reassign": function noNativeReassign() {
    return require("./no-native-reassign");
  },
  "no-negated-condition": function noNegatedCondition() {
    return require("./no-negated-condition");
  },
  "no-negated-in-lhs": function noNegatedInLhs() {
    return require("./no-negated-in-lhs");
  },
  "no-nested-ternary": function noNestedTernary() {
    return require("./no-nested-ternary");
  },
  "no-new": function noNew() {
    return require("./no-new");
  },
  "no-new-func": function noNewFunc() {
    return require("./no-new-func");
  },
  "no-new-object": function noNewObject() {
    return require("./no-new-object");
  },
  "no-new-require": function noNewRequire() {
    return require("./no-new-require");
  },
  "no-new-symbol": function noNewSymbol() {
    return require("./no-new-symbol");
  },
  "no-new-wrappers": function noNewWrappers() {
    return require("./no-new-wrappers");
  },
  "no-obj-calls": function noObjCalls() {
    return require("./no-obj-calls");
  },
  "no-octal": function noOctal() {
    return require("./no-octal");
  },
  "no-octal-escape": function noOctalEscape() {
    return require("./no-octal-escape");
  },
  "no-param-reassign": function noParamReassign() {
    return require("./no-param-reassign");
  },
  "no-path-concat": function noPathConcat() {
    return require("./no-path-concat");
  },
  "no-plusplus": function noPlusplus() {
    return require("./no-plusplus");
  },
  "no-process-env": function noProcessEnv() {
    return require("./no-process-env");
  },
  "no-process-exit": function noProcessExit() {
    return require("./no-process-exit");
  },
  "no-proto": function noProto() {
    return require("./no-proto");
  },
  "no-prototype-builtins": function noPrototypeBuiltins() {
    return require("./no-prototype-builtins");
  },
  "no-redeclare": function noRedeclare() {
    return require("./no-redeclare");
  },
  "no-regex-spaces": function noRegexSpaces() {
    return require("./no-regex-spaces");
  },
  "no-restricted-globals": function noRestrictedGlobals() {
    return require("./no-restricted-globals");
  },
  "no-restricted-imports": function noRestrictedImports() {
    return require("./no-restricted-imports");
  },
  "no-restricted-modules": function noRestrictedModules() {
    return require("./no-restricted-modules");
  },
  "no-restricted-properties": function noRestrictedProperties() {
    return require("./no-restricted-properties");
  },
  "no-restricted-syntax": function noRestrictedSyntax() {
    return require("./no-restricted-syntax");
  },
  "no-return-assign": function noReturnAssign() {
    return require("./no-return-assign");
  },
  "no-return-await": function noReturnAwait() {
    return require("./no-return-await");
  },
  "no-script-url": function noScriptUrl() {
    return require("./no-script-url");
  },
  "no-self-assign": function noSelfAssign() {
    return require("./no-self-assign");
  },
  "no-self-compare": function noSelfCompare() {
    return require("./no-self-compare");
  },
  "no-sequences": function noSequences() {
    return require("./no-sequences");
  },
  "no-setter-return": function noSetterReturn() {
    return require("./no-setter-return");
  },
  "no-shadow": function noShadow() {
    return require("./no-shadow");
  },
  "no-shadow-restricted-names": function noShadowRestrictedNames() {
    return require("./no-shadow-restricted-names");
  },
  "no-spaced-func": function noSpacedFunc() {
    return require("./no-spaced-func");
  },
  "no-sparse-arrays": function noSparseArrays() {
    return require("./no-sparse-arrays");
  },
  "no-sync": function noSync() {
    return require("./no-sync");
  },
  "no-tabs": function noTabs() {
    return require("./no-tabs");
  },
  "no-template-curly-in-string": function noTemplateCurlyInString() {
    return require("./no-template-curly-in-string");
  },
  "no-ternary": function noTernary() {
    return require("./no-ternary");
  },
  "no-this-before-super": function noThisBeforeSuper() {
    return require("./no-this-before-super");
  },
  "no-throw-literal": function noThrowLiteral() {
    return require("./no-throw-literal");
  },
  "no-trailing-spaces": function noTrailingSpaces() {
    return require("./no-trailing-spaces");
  },
  "no-undef": function noUndef() {
    return require("./no-undef");
  },
  "no-undef-init": function noUndefInit() {
    return require("./no-undef-init");
  },
  "no-undefined": function noUndefined() {
    return require("./no-undefined");
  },
  "no-underscore-dangle": function noUnderscoreDangle() {
    return require("./no-underscore-dangle");
  },
  "no-unexpected-multiline": function noUnexpectedMultiline() {
    return require("./no-unexpected-multiline");
  },
  "no-unmodified-loop-condition": function noUnmodifiedLoopCondition() {
    return require("./no-unmodified-loop-condition");
  },
  "no-unneeded-ternary": function noUnneededTernary() {
    return require("./no-unneeded-ternary");
  },
  "no-unreachable": function noUnreachable() {
    return require("./no-unreachable");
  },
  "no-unsafe-finally": function noUnsafeFinally() {
    return require("./no-unsafe-finally");
  },
  "no-unsafe-negation": function noUnsafeNegation() {
    return require("./no-unsafe-negation");
  },
  "no-unused-expressions": function noUnusedExpressions() {
    return require("./no-unused-expressions");
  },
  "no-unused-labels": function noUnusedLabels() {
    return require("./no-unused-labels");
  },
  "no-unused-vars": function noUnusedVars() {
    return require("./no-unused-vars");
  },
  "no-use-before-define": function noUseBeforeDefine() {
    return require("./no-use-before-define");
  },
  "no-useless-call": function noUselessCall() {
    return require("./no-useless-call");
  },
  "no-useless-catch": function noUselessCatch() {
    return require("./no-useless-catch");
  },
  "no-useless-computed-key": function noUselessComputedKey() {
    return require("./no-useless-computed-key");
  },
  "no-useless-concat": function noUselessConcat() {
    return require("./no-useless-concat");
  },
  "no-useless-constructor": function noUselessConstructor() {
    return require("./no-useless-constructor");
  },
  "no-useless-escape": function noUselessEscape() {
    return require("./no-useless-escape");
  },
  "no-useless-rename": function noUselessRename() {
    return require("./no-useless-rename");
  },
  "no-useless-return": function noUselessReturn() {
    return require("./no-useless-return");
  },
  "no-var": function noVar() {
    return require("./no-var");
  },
  "no-void": function noVoid() {
    return require("./no-void");
  },
  "no-warning-comments": function noWarningComments() {
    return require("./no-warning-comments");
  },
  "no-whitespace-before-property": function noWhitespaceBeforeProperty() {
    return require("./no-whitespace-before-property");
  },
  "no-with": function noWith() {
    return require("./no-with");
  },
  "nonblock-statement-body-position": function nonblockStatementBodyPosition() {
    return require("./nonblock-statement-body-position");
  },
  "object-curly-newline": function objectCurlyNewline() {
    return require("./object-curly-newline");
  },
  "object-curly-spacing": function objectCurlySpacing() {
    return require("./object-curly-spacing");
  },
  "object-property-newline": function objectPropertyNewline() {
    return require("./object-property-newline");
  },
  "object-shorthand": function objectShorthand() {
    return require("./object-shorthand");
  },
  "one-var": function oneVar() {
    return require("./one-var");
  },
  "one-var-declaration-per-line": function oneVarDeclarationPerLine() {
    return require("./one-var-declaration-per-line");
  },
  "operator-assignment": function operatorAssignment() {
    return require("./operator-assignment");
  },
  "operator-linebreak": function operatorLinebreak() {
    return require("./operator-linebreak");
  },
  "padded-blocks": function paddedBlocks() {
    return require("./padded-blocks");
  },
  "padding-line-between-statements": function paddingLineBetweenStatements() {
    return require("./padding-line-between-statements");
  },
  "prefer-arrow-callback": function preferArrowCallback() {
    return require("./prefer-arrow-callback");
  },
  "prefer-const": function preferConst() {
    return require("./prefer-const");
  },
  "prefer-destructuring": function preferDestructuring() {
    return require("./prefer-destructuring");
  },
  "prefer-exponentiation-operator": function preferExponentiationOperator() {
    return require("./prefer-exponentiation-operator");
  },
  "prefer-named-capture-group": function preferNamedCaptureGroup() {
    return require("./prefer-named-capture-group");
  },
  "prefer-numeric-literals": function preferNumericLiterals() {
    return require("./prefer-numeric-literals");
  },
  "prefer-object-spread": function preferObjectSpread() {
    return require("./prefer-object-spread");
  },
  "prefer-promise-reject-errors": function preferPromiseRejectErrors() {
    return require("./prefer-promise-reject-errors");
  },
  "prefer-reflect": function preferReflect() {
    return require("./prefer-reflect");
  },
  "prefer-regex-literals": function preferRegexLiterals() {
    return require("./prefer-regex-literals");
  },
  "prefer-rest-params": function preferRestParams() {
    return require("./prefer-rest-params");
  },
  "prefer-spread": function preferSpread() {
    return require("./prefer-spread");
  },
  "prefer-template": function preferTemplate() {
    return require("./prefer-template");
  },
  "quote-props": function quoteProps() {
    return require("./quote-props");
  },
  quotes: function quotes() {
    return require("./quotes");
  },
  radix: function radix() {
    return require("./radix");
  },
  "require-atomic-updates": function requireAtomicUpdates() {
    return require("./require-atomic-updates");
  },
  "require-await": function requireAwait() {
    return require("./require-await");
  },
  "require-jsdoc": function requireJsdoc() {
    return require("./require-jsdoc");
  },
  "require-unicode-regexp": function requireUnicodeRegexp() {
    return require("./require-unicode-regexp");
  },
  "require-yield": function requireYield() {
    return require("./require-yield");
  },
  "rest-spread-spacing": function restSpreadSpacing() {
    return require("./rest-spread-spacing");
  },
  semi: function semi() {
    return require("./semi");
  },
  "semi-spacing": function semiSpacing() {
    return require("./semi-spacing");
  },
  "semi-style": function semiStyle() {
    return require("./semi-style");
  },
  "sort-imports": function sortImports() {
    return require("./sort-imports");
  },
  "sort-keys": function sortKeys() {
    return require("./sort-keys");
  },
  "sort-vars": function sortVars() {
    return require("./sort-vars");
  },
  "space-before-blocks": function spaceBeforeBlocks() {
    return require("./space-before-blocks");
  },
  "space-before-function-paren": function spaceBeforeFunctionParen() {
    return require("./space-before-function-paren");
  },
  "space-in-parens": function spaceInParens() {
    return require("./space-in-parens");
  },
  "space-infix-ops": function spaceInfixOps() {
    return require("./space-infix-ops");
  },
  "space-unary-ops": function spaceUnaryOps() {
    return require("./space-unary-ops");
  },
  "spaced-comment": function spacedComment() {
    return require("./spaced-comment");
  },
  strict: function strict() {
    return require("./strict");
  },
  "switch-colon-spacing": function switchColonSpacing() {
    return require("./switch-colon-spacing");
  },
  "symbol-description": function symbolDescription() {
    return require("./symbol-description");
  },
  "template-curly-spacing": function templateCurlySpacing() {
    return require("./template-curly-spacing");
  },
  "template-tag-spacing": function templateTagSpacing() {
    return require("./template-tag-spacing");
  },
  "unicode-bom": function unicodeBom() {
    return require("./unicode-bom");
  },
  "use-isnan": function useIsnan() {
    return require("./use-isnan");
  },
  "valid-jsdoc": function validJsdoc() {
    return require("./valid-jsdoc");
  },
  "valid-typeof": function validTypeof() {
    return require("./valid-typeof");
  },
  "vars-on-top": function varsOnTop() {
    return require("./vars-on-top");
  },
  "wrap-iife": function wrapIife() {
    return require("./wrap-iife");
  },
  "wrap-regex": function wrapRegex() {
    return require("./wrap-regex");
  },
  "yield-star-spacing": function yieldStarSpacing() {
    return require("./yield-star-spacing");
  },
  yoda: function yoda() {
    return require("./yoda");
  }
}));
