// eslint/lib/rules/dist/max-classes-per-file.dev.js
/**
 * @fileoverview Enforce a maximum number of classes per file
 * @author James Garbutt <https://github.com/43081j>
 */
"use strict"; //------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
  meta: {
    type: "suggestion",
    docs: {
      description: "enforce a maximum number of classes per file",
      category: "Best Practices",
      recommended: false,
      url: "https://eslint.org/docs/rules/max-classes-per-file"
    },
    schema: [{
      type: "integer",
      minimum: 1
    }],
    messages: {
      maximumExceeded: "File has too many classes ({{ classCount }}). Maximum allowed is {{ max }}."
    }
  },
  create: function create(context) {
    var maxClasses = context.options[0] || 1;
    var classCount = 0;
    return {
      Program: function Program() {
        classCount = 0;
      },
      "Program:exit": function ProgramExit(node) {
        if (classCount > maxClasses) {
          context.report({
            node: node,
            messageId: "maximumExceeded",
            data: {
              classCount: classCount,
              max: maxClasses
            }
          });
        }
      },
      "ClassDeclaration, ClassExpression": function ClassDeclarationClassExpression() {
        classCount++;
      }
    };
  }
};
