// eslint/lib/rules/dist/no-async-promise-executor.dev.js
/**
 * @fileoverview disallow using an async function as a Promise executor
 * @author Teddy Katz
 */
"use strict"; //------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
  meta: {
    type: "problem",
    docs: {
      description: "disallow using an async function as a Promise executor",
      category: "Possible Errors",
      recommended: true,
      url: "https://eslint.org/docs/rules/no-async-promise-executor"
    },
    fixable: null,
    schema: [],
    messages: {
      async: "Promise executor functions should not be async."
    }
  },
  create: function create(context) {
    return {
      "NewExpression[callee.name='Promise'][arguments.0.async=true]": function NewExpressionCalleeNamePromiseArguments0AsyncTrue(node) {
        context.report({
          node: context.getSourceCode().getFirstToken(node.arguments[0], function (token) {
            return token.value === "async";
          }),
          messageId: "async"
        });
      }
    };
  }
};
