// eslint/lib/rules/dist/no-constructor-return.dev.js
/**
 * @fileoverview Rule to disallow returning value from constructor.
 * @author Pig Fang <https://github.com/g-plane>
 */
"use strict"; //------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
  meta: {
    type: "problem",
    docs: {
      description: "disallow returning value from constructor",
      category: "Best Practices",
      recommended: false,
      url: "https://eslint.org/docs/rules/no-constructor-return"
    },
    schema: {},
    fixable: null,
    messages: {
      unexpected: "Unexpected return statement in constructor."
    }
  },
  create: function create(context) {
    var stack = [];
    return {
      onCodePathStart: function onCodePathStart(_, node) {
        stack.push(node);
      },
      onCodePathEnd: function onCodePathEnd() {
        stack.pop();
      },
      ReturnStatement: function ReturnStatement(node) {
        var last = stack[stack.length - 1];

        if (!last.parent) {
          return;
        }

        if (last.parent.type === "MethodDefinition" && last.parent.kind === "constructor" && (node.parent.parent === last || node.argument)) {
          context.report({
            node: node,
            messageId: "unexpected"
          });
        }
      }
    };
  }
};
