// eslint/lib/rules/dist/no-multi-assign.dev.js
/**
 * @fileoverview Rule to check use of chained assignment expressions
 * @author Stewart Rand
 */
"use strict"; //------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
  meta: {
    type: "suggestion",
    docs: {
      description: "disallow use of chained assignment expressions",
      category: "Stylistic Issues",
      recommended: false,
      url: "https://eslint.org/docs/rules/no-multi-assign"
    },
    schema: []
  },
  create: function create(context) {
    //--------------------------------------------------------------------------
    // Public
    //--------------------------------------------------------------------------
    return {
      AssignmentExpression: function AssignmentExpression(node) {
        if (["AssignmentExpression", "VariableDeclarator"].indexOf(node.parent.type) !== -1) {
          context.report({
            node: node,
            message: "Unexpected chained assignment."
          });
        }
      }
    };
  }
};
