// eslint/lib/rules/dist/no-throw-literal.dev.js
/**
 * @fileoverview Rule to restrict what can be thrown as an exception.
 * @author Dieter Oberkofler
 */
"use strict";

var astUtils = require("./utils/ast-utils"); //------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------


module.exports = {
  meta: {
    type: "suggestion",
    docs: {
      description: "disallow throwing literals as exceptions",
      category: "Best Practices",
      recommended: false,
      url: "https://eslint.org/docs/rules/no-throw-literal"
    },
    schema: [],
    messages: {
      object: "Expected an error object to be thrown.",
      undef: "Do not throw undefined."
    }
  },
  create: function create(context) {
    return {
      ThrowStatement: function ThrowStatement(node) {
        if (!astUtils.couldBeError(node.argument)) {
          context.report({
            node: node,
            messageId: "object"
          });
        } else if (node.argument.type === "Identifier") {
          if (node.argument.name === "undefined") {
            context.report({
              node: node,
              messageId: "undef"
            });
          }
        }
      }
    };
  }
};
