// eslint/lib/rules/dist/no-template-curly-in-string.dev.js
/**
 * @fileoverview Warn when using template string syntax in regular strings
 * @author Jeroen Engels
 */
"use strict"; //------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
  meta: {
    type: "problem",
    docs: {
      description: "disallow template literal placeholder syntax in regular strings",
      category: "Possible Errors",
      recommended: false,
      url: "https://eslint.org/docs/rules/no-template-curly-in-string"
    },
    schema: []
  },
  create: function create(context) {
    var regex = /\$\{(?:[\0-\|~-\uD7FF\uE000-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])+\}/;
    return {
      Literal: function Literal(node) {
        if (typeof node.value === "string" && regex.test(node.value)) {
          context.report({
            node: node,
            message: "Unexpected template string expression."
          });
        }
      }
    };
  }
};
