// eslint/lib/rules/dist/no-void.dev.js
/**
 * @fileoverview Rule to disallow use of void operator.
 * @author Mike Sidorov
 */
"use strict"; //------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
  meta: {
    type: "suggestion",
    docs: {
      description: "disallow `void` operators",
      category: "Best Practices",
      recommended: false,
      url: "https://eslint.org/docs/rules/no-void"
    },
    schema: []
  },
  create: function create(context) {
    //--------------------------------------------------------------------------
    // Public
    //--------------------------------------------------------------------------
    return {
      UnaryExpression: function UnaryExpression(node) {
        if (node.operator === "void") {
          context.report({
            node: node,
            message: "Expected 'undefined' and instead saw 'void'."
          });
        }
      }
    };
  }
};
