// @typescript-eslint/eslint-plugin/dist/rules/dist/no-extra-semi.dev.js
"use strict";

var __importDefault = void 0 && (void 0).__importDefault || function (mod) {
  return mod && mod.__esModule ? mod : {
    "default": mod
  };
};

var __importStar = void 0 && (void 0).__importStar || function (mod) {
  if (mod && mod.__esModule) return mod;
  var result = {};
  if (mod != null) for (var k in mod) {
    if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
  }
  result["default"] = mod;
  return result;
};

Object.defineProperty(exports, "__esModule", {
  value: true
});

var no_extra_semi_1 = __importDefault(require("eslint/lib/rules/no-extra-semi"));

var util = __importStar(require("../util"));

exports["default"] = util.createRule({
  name: 'no-extra-semi',
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Disallow unnecessary semicolons',
      category: 'Possible Errors',
      recommended: false,
      extendsBaseRule: true
    },
    fixable: 'code',
    schema: no_extra_semi_1["default"].meta.schema,
    messages: no_extra_semi_1["default"].meta.messages
  },
  defaultOptions: [],
  create: function create(context) {
    var rules = no_extra_semi_1["default"].create(context);
    return Object.assign(Object.assign({}, rules), {
      ClassProperty: function ClassProperty(node) {
        rules.MethodDefinition(node);
      }
    });
  }
});
