// @babel/preset-react/lib/dist/index.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _helperPluginUtils = require("@babel/helper-plugin-utils");

var _pluginTransformReactJsx = _interopRequireDefault(require("@babel/plugin-transform-react-jsx"));

var _pluginTransformReactJsxDevelopment = _interopRequireDefault(require("@babel/plugin-transform-react-jsx-development"));

var _pluginTransformReactDisplayName = _interopRequireDefault(require("@babel/plugin-transform-react-display-name"));

var _pluginTransformReactJsxSource = _interopRequireDefault(require("@babel/plugin-transform-react-jsx-source"));

var _pluginTransformReactJsxSelf = _interopRequireDefault(require("@babel/plugin-transform-react-jsx-self"));

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    "default": obj
  };
}

var _default = (0, _helperPluginUtils.declare)(function (api, opts) {
  api.assertVersion(7);
  var pragma = opts.pragma,
      pragmaFrag = opts.pragmaFrag,
      pure = opts.pure,
      _opts$throwIfNamespac = opts.throwIfNamespace,
      throwIfNamespace = _opts$throwIfNamespac === void 0 ? true : _opts$throwIfNamespac,
      useSpread = opts.useSpread,
      _opts$runtime = opts.runtime,
      runtime = _opts$runtime === void 0 ? "classic" : _opts$runtime,
      importSource = opts.importSource;
  var development = !!opts.development;
  var useBuiltIns = !!opts.useBuiltIns;

  if (typeof development !== "boolean") {
    throw new Error("@babel/preset-react 'development' option must be a boolean.");
  }

  var transformReactJSXPlugin = runtime === "automatic" && development ? _pluginTransformReactJsxDevelopment["default"] : _pluginTransformReactJsx["default"];
  return {
    plugins: [[transformReactJSXPlugin, {
      importSource: importSource,
      pragma: pragma,
      pragmaFrag: pragmaFrag,
      runtime: runtime,
      throwIfNamespace: throwIfNamespace,
      useBuiltIns: useBuiltIns,
      useSpread: useSpread,
      pure: pure
    }], _pluginTransformReactDisplayName["default"], development && runtime === "classic" && _pluginTransformReactJsxSource["default"], development && runtime === "classic" && _pluginTransformReactJsxSelf["default"]].filter(Boolean)
  };
});

exports["default"] = _default;
