// default-gateway/dist/android.dev.js
"use strict";

var net = require("net");

var execa = require("execa");

var args = {
  v4: ["-4", "r"],
  v6: ["-6", "r"]
};

var parse = function parse(stdout) {
  var result;
  (stdout || "").trim().split("\n").some(function (line) {
    var results = /default via (.+?) dev (.+?)( |$)/.exec(line) || [];
    var gateway = results[1];
    var iface = results[2];

    if (gateway && net.isIP(gateway)) {
      result = {
        gateway: gateway,
        "interface": iface ? iface : null
      };
      return true;
    }
  });

  if (!result) {
    throw new Error("Unable to determine default gateway");
  }

  return result;
};

var promise = function promise(family) {
  return execa.stdout("ip", args[family]).then(function (stdout) {
    return parse(stdout);
  });
};

var sync = function sync(family) {
  var result = execa.sync("ip", args[family]);
  return parse(result.stdout);
};

module.exports.v4 = function () {
  return promise("v4");
};

module.exports.v6 = function () {
  return promise("v6");
};

module.exports.v4.sync = function () {
  return sync("v4");
};

module.exports.v6.sync = function () {
  return sync("v6");
};
