// default-gateway/dist/darwin.dev.js
"use strict";

var net = require("net");

var execa = require("execa");

var dests = ["default", "0.0.0.0", "0.0.0.0/0", "::", "::/0"];
var args = {
  v4: ["-rn", "-f", "inet"],
  v6: ["-rn", "-f", "inet6"]
};

var parse = function parse(stdout, family) {
  var result;
  (stdout || "").trim().split("\n").some(function (line) {
    var results = line.split(/ +/) || [];
    var target = results[0];
    var gateway = results[1];
    var iface = results[family === "v4" ? 5 : 3];

    if (dests.indexOf(target) !== -1 && gateway && net.isIP(gateway)) {
      result = {
        gateway: gateway,
        "interface": iface ? iface : null
      };
      return true;
    }
  });

  if (!result) {
    throw new Error("Unable to determine default gateway");
  }

  return result;
};

var promise = function promise(family) {
  return execa.stdout("netstat", args[family]).then(function (stdout) {
    return parse(stdout, family);
  });
};

var sync = function sync(family) {
  var result = execa.sync("netstat", args[family]);
  return parse(result.stdout, family);
};

module.exports.v4 = function () {
  return promise("v4");
};

module.exports.v6 = function () {
  return promise("v6");
};

module.exports.v4.sync = function () {
  return sync("v4");
};

module.exports.v6.sync = function () {
  return sync("v6");
};
