// @typescript-eslint/experimental-utils/dist/ts-eslint-scope/dist/Scope.dev.js
"use strict";
/* eslint-disable @typescript-eslint/no-empty-interface, @typescript-eslint/no-explicit-any */

Object.defineProperty(exports, "__esModule", {
  value: true
});

var scope_1 = require("eslint-scope/lib/scope");

var Scope = scope_1.Scope;
exports.Scope = Scope;
var GlobalScope = scope_1.GlobalScope;
exports.GlobalScope = GlobalScope;
var ModuleScope = scope_1.ModuleScope;
exports.ModuleScope = ModuleScope;
var FunctionExpressionNameScope = scope_1.FunctionExpressionNameScope;
exports.FunctionExpressionNameScope = FunctionExpressionNameScope;
var CatchScope = scope_1.CatchScope;
exports.CatchScope = CatchScope;
var WithScope = scope_1.WithScope;
exports.WithScope = WithScope;
var BlockScope = scope_1.BlockScope;
exports.BlockScope = BlockScope;
var SwitchScope = scope_1.SwitchScope;
exports.SwitchScope = SwitchScope;
var FunctionScope = scope_1.FunctionScope;
exports.FunctionScope = FunctionScope;
var ForScope = scope_1.ForScope;
exports.ForScope = ForScope;
var ClassScope = scope_1.ClassScope;
exports.ClassScope = ClassScope;
