// @svgr/babel-plugin-transform-svg-component/lib/dist/index.dev.js
"use strict";

function _templateObject() {
  var data = _taggedTemplateLiteral(["", "\nconst ", " = (", ") => ", "\n", "\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

exports.__esModule = true;
exports["default"] = void 0;

var _util = require("./util");

function defaultTemplate(_ref, opts, _ref2) {
  var template = _ref.template;
  var imports = _ref2.imports,
      componentName = _ref2.componentName,
      props = _ref2.props,
      jsx = _ref2.jsx,
      exports = _ref2.exports;
  return template.ast(_templateObject(), imports, componentName, props, jsx, exports);
}

var plugin = function plugin(api, opts) {
  return {
    visitor: {
      Program: function Program(path) {
        var t = api.types;
        var template = opts.template || defaultTemplate;
        var body = template(api, opts, {
          componentName: t.identifier(opts.state.componentName),
          props: (0, _util.getProps)(api, opts),
          imports: (0, _util.getImport)(api, opts),
          exports: (0, _util.getExport)(api, opts),
          jsx: path.node.body[0].expression
        });

        if (Array.isArray(body)) {
          path.node.body = body;
        } else {
          path.node.body = [body];
        }

        path.replaceWith(path.node);
      }
    }
  };
};

var _default = plugin;
exports["default"] = _default;
